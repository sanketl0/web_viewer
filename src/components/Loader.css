.gvs .gvs-loader {
	position: relative;
	width: 100%;
	height: 100%;
	z-index: 0;
}

.gvs .gvs-loader.gvs-loader-visible {
	visibility: visible;
	opacity: 1;
}

/* Loader overlay */
.gvs .gvs-loader {
	visibility: hidden;
	position: absolute;
	top: 0;
	right: 0;
	left: 0;
	bottom: 0;
	opacity: 0;
	display: flex;
	flex-direction: column;
	justify-content: center;
	gap: 20px;
	align-items: center;
	background: #37474F;
	z-index: 9000;
	transition: all 0.5s;
	font-family: 'Gill Sans', 'Gill Sans MT', Calibri, 'Trebuchet MS', sans-serif;
	font-weight: 550;
	color: white;
	font-size: 1.4em;
	text-align: center;
}

.gvs .gvs-loader a {
	color: white !important;
}

/* Flashing text */
.gvs .gvs-loader > div > span {
	animation: blinker 2s linear infinite;
}
@keyframes blinker { 50% { opacity: 0.3; } }

/* Rotating logo */
.gvs .gvs-loader .gvs-loader-img {
	width: 150px;
	animation: rotating 1.4s linear infinite;
}

@keyframes rotating {
	from { transform: rotate(0deg); }
	to { transform: rotate(360deg); }
}

/* Call to action button */
.gvs .gvs-loader .gvs-loader-cta {
	border: none;
	border-radius: 7px;
	font-weight: 500;
	font-size: 1.0em;
	color: white;
	background-color: #EF6C00;
	cursor: pointer;
	margin: 10px;
	padding: 5px 30px;
	box-shadow: 2px 2px 2px rgba(0,0,0,0.3);
}

.gvs .gvs-loader .gvs-loader-cta:hover {
	background-color: #F57C00;
}