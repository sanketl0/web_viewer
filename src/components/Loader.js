import "./Loader.css";
import LogoDead from "../img/logo_dead.svg";
import LoaderImg from "../img/marker.svg";

/**
 * Loader is a full-screen loading message.
 * @private
 * 
 * @param {CoreView} parent The parent view
 * @param {Element} container The DOM element to create loader into
 */
export default class Loader {
	constructor(parent, container) {
		this._parent = parent;
		this.container = container;
		this.container.classList.add("gvs-loader", "gvs-loader-visible");

		// Logo
		const logo = document.createElement("img");
		logo.src = LoaderImg;
		logo.alt = "";
		logo.title = this._parent._t.map.loading;
		logo.classList.add("gvs-loader-img");
		this.container.appendChild(logo);

		// Label (1 serious, then fun ones)
		const labelWrapper = document.createElement("div");
		const label = document.createElement("span");
		const nextLabelFun = () => (
			this._parent._t.gvs.loading_labels_fun[
				Math.floor(Math.random() * this._parent._t.gvs.loading_labels_fun.length)
			]
		);
		label.innerHTML = this._parent._t.gvs.loading_labels_serious[
			Math.floor(Math.random() * this._parent._t.gvs.loading_labels_serious.length)
		];
		const nextLabelFct = () => setTimeout(() => {
			label.innerHTML = nextLabelFun();
			this._loaderLabelChanger = nextLabelFct();
		}, 500 + Math.random() * 1000);
		this._loaderLabelChanger = nextLabelFct();
		labelWrapper.appendChild(label);

		this.container.appendChild(labelWrapper);
	}

	/**
	 * Is the loader visible ?
	 * @returns {boolean} True if visible
	 */
	isVisible() {
		return this.container.classList.contains("gvs-loader-visible");
	}

	/**
	 * Dismiss loader, or show error
	 * @param {object} [err] Optional error object to show in browser console
	 * @param {str} [errMeaningful] Optional error message to show to user
	 * @param {fct} [next] Optional function to run after loader dismiss
	 */
	dismiss(err = null, errMeaningful = null, next = null) {
		clearTimeout(this._loaderLabelChanger);

		if(!err) {
			this.container.classList.remove("gvs-loader-visible");
			setTimeout(() => this.container.remove(), 2000);

			/**
			 * Event for viewer being ready to use (API loaded)
			 *
			 * @event ready
			 * @memberof CoreView
			 */
			const readyEvt = new CustomEvent("ready");
			this._parent.dispatchEvent(readyEvt);

			if(next) { next(); }
		}
		else {
			if(err !== true) { console.error(err); }

			// Change content
			this.container.children[0].src = LogoDead;
			this.container.children[0].style.width = "200px";
			this.container.children[0].style.animation = "unset";

			let errHtml;
			if(next) {
				errHtml = `<button class="gvs-loader-cta">${this._parent._t.gvs.error_click}</button>`;
			}
			else {
				errHtml = `<small>${this._parent._t.gvs.error_retry}</small>`;
			}

			if(errMeaningful) { errHtml = errMeaningful + "<br />" + errHtml; }
			this.container.children[1].innerHTML = `${this._parent._t.gvs.error}<br />${errHtml}`;
			if(next) {
				this.container.addEventListener("click", next);
			}
			const errLabel = errMeaningful || "Panoramax JS had a blocking exception";

			/**
			 * Event for viewer failing to initially load
			 *
			 * @event broken
			 * @memberof CoreView
			 * @type {object}
			 * @property {object} detail Event information
			 * @property {string} detail.error The user-friendly error message to display
			 */
			const brokenEvt = new CustomEvent("broken", {
				detail: { error: errLabel }
			});
			this._parent.dispatchEvent(brokenEvt);

			// Throw error
			throw new Error(errLabel);
		}
	}
}