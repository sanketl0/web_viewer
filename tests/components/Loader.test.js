import Loader from "../../src/components/Loader";

describe("constructor", () => {
	it("works", () => {
		const p = { _t: { gvs: { loading_labels_serious: ["Loading..."], loading_labels_fun: ["FUN..."] }, map: {} } };
		const c = document.createElement("div");
		const l = new Loader(p, c);
		expect(l.container.className).toBe("gvs-loader gvs-loader-visible");
		expect(l.container.innerHTML).toMatchSnapshot();
	});
});

describe("dismiss", () => {
	it("works when no error set", () => {
		const p = {
			_t: { gvs: { loading_labels_serious: ["Loading..."], loading_labels_fun: ["FUN..."] }, map: {} },
			dispatchEvent: jest.fn()
		};
		const c = document.createElement("div");
		const l = new Loader(p, c);
		l.dismiss();
		expect(c.className).toBe("gvs-loader");
		expect(p.dispatchEvent.mock.calls).toMatchSnapshot();
	});

	it("works with error set", () => {
		global.console = { error: jest.fn() };
		const p = {
			_t: { gvs: { loading_labels_serious: ["Loading..."], loading_labels_fun: ["FUN..."] }, map: {} },
			dispatchEvent: jest.fn()
		};
		const c = document.createElement("div");
		const l = new Loader(p, c);
		expect(() => l.dismiss("Technical issue", "Oops it's broken")).toThrowError("Oops it's broken");
		expect(c.className).toBe("gvs-loader gvs-loader-visible");
		expect(c.innerHTML).toMatchSnapshot();
	});
});