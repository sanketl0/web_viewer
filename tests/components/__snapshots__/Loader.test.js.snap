// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`constructor works 1`] = `"<img src=\\"marker.svg\\" alt=\\"\\" title=\\"undefined\\" class=\\"gvs-loader-img\\"><div><span>Loading...</span></div>"`;

exports[`dismiss works when no error set 1`] = `
Array [
  Array [
    CustomEvent {
      "isTrusted": false,
    },
  ],
]
`;

exports[`dismiss works with error set 1`] = `"<img src=\\"logo_dead.svg\\" alt=\\"\\" title=\\"undefined\\" class=\\"gvs-loader-img\\" style=\\"width: 200px; animation: unset;\\"><div>undefined<br>Oops it's broken<br><small>undefined</small></div>"`;
